yarn-project/end-to-end/src/e2e_key_registry.test.ts

```diff
@@ -0,0 +1,252 @@
+import { type AccountWallet, AztecAddress, Fr, type PXE } from '@aztec/aztec.js';
+import { GeneratorIndex } from '@aztec/circuits.js';
+import { poseidon2Hash } from '@aztec/foundation/crypto';
+import { KeyRegistryContract, StateVarsContract } from '@aztec/noir-contracts.js';
+
+import { jest } from '@jest/globals';
+
+import { publicDeployAccounts, setup } from './fixtures/utils.js';
+
+const TIMEOUT = 100_000;
+
+describe('SharedMutablePrivateGetter', () => {
+  let keyRegistry: KeyRegistryContract;
+  let stateVarsContract: StateVarsContract;
+  let pxe: PXE;
+  jest.setTimeout(TIMEOUT);
+
+  let wallets: AccountWallet[];
+
+  let teardown: () => Promise<void>;
+
+  beforeAll(async () => {
+    ({ teardown, pxe, wallets } = await setup(2));
+    stateVarsContract = await StateVarsContract.deploy(wallets[0]).send().deployed();
+    keyRegistry = await KeyRegistryContract.deploy(wallets[0]).send().deployed();
+
+    await publicDeployAccounts(wallets[0], wallets.slice(0, 2));
+  }, 120_000);
+
+  const delay = async (blocks: number) => {
+    for (let i = 0; i < blocks; i++) {
+      await stateVarsContract.methods.delay().send().wait();
+    }
+  };
+
+  afterAll(() => teardown());
+
+  describe('failure cases', () => {
+    let accountAddedToRegistry: AztecAddress;
+
+    describe('should fail registering or rotating with bad input', () => {
+      const partialAddress = new Fr(69);
+
+      const masterNullifierPublicKey = new Fr(12);
+      const masterIncomingViewingPublicKey = new Fr(34);
+      const masterOutgoingViewingPublicKey = new Fr(56);
+      const masterTaggingPublicKey = new Fr(78);
+
+      const publicKeysHash = poseidon2Hash([
+        masterNullifierPublicKey,
+        masterIncomingViewingPublicKey,
+        masterOutgoingViewingPublicKey,
+        masterTaggingPublicKey,
+        GeneratorIndex.PUBLIC_KEYS_HASH,
+      ]);
+
+      // We hash the partial address and the public keys hash to get the account address
+      // TODO(#5726): Should GeneratorIndex.CONTRACT_ADDRESS be removed given that we introduced CONTRACT_ADDRESS_V1?
+      // TODO(#5726): Move the following line to AztecAddress class?
+      accountAddedToRegistry = poseidon2Hash([partialAddress, publicKeysHash, GeneratorIndex.CONTRACT_ADDRESS_V1]);
+
+      it('should fail registering with mismatched address', async () => {
+        const mismatchedAddress = accountAddedToRegistry.add(new Fr(1));
```

Comment:
```suggestion
        const mismatchedAddress = Fr.random();
```
Just using random value here is clearer to read.